### PR TITLE
ci: Update pipelines to verify compilation on Java 17

### PR DESCRIPTION
**Pull request by manusa**
## Description
ci: Update pipelines to verify compilation on Java 17

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by manusa** [comment]
Unsupported in Karaf

```
2021-11-24T16:03:48,909 | ERROR | features-3-thread-1 | ManagedKubernetesClient          | 25 - io.fabric8.kubernetes-client - 5.11.0.SNAPSHOT | bundle io.fabric8.kubernetes-client:5.11.0.SNAPSHOT (25)[io.fabric8.kubernetes.client.osgi.ManagedKubernetesClient(1)] : Error during instantiation of the implementation object
java.lang.reflect.InvocationTargetException: null
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
	at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:316) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:286) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:776) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437) ~[?:?]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:667) ~[?:?]
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:305) ~[?:?]
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:554) ~[?:?]
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:70) ~[?:?]
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:421) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49) ~[?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:420) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450) ~[osgi.core-7.0.0.jar:?]
	at org.apache.felix.framework.EventDispatcher.invokeBundleListenerCallback(EventDispatcher.java:915) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:834) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireBundleEvent(EventDispatcher.java:516) ~[?:?]
	at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4817) ~[?:?]
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2336) ~[?:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) ~[?:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) ~[?:?]
	at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.startBundle(BundleInstallSupportImpl.java:165) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1160) ~[?:?]
	at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1041) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.lang.ClassFormatError: Weaving hook failed.
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.transformClass(BundleWiringImpl.java:2371) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.findClass(BundleWiringImpl.java:2049) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1533) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:78) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1950) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:11) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:8) ~[?:?]
	at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:45) ~[?:?]
	at io.fabric8.kubernetes.client.osgi.ManagedKubernetesClient.<init>(ManagedKubernetesClient.java:137) ~[?:?]
	... 41 more
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 61
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:196) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:177) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:163) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:284) ~[?:?]
	at org.apache.aries.spifly.dynamic.OSGiFriendlyClassWriter.getCommonSuperClass(OSGiFriendlyClassWriter.java:84) ~[?:?]
	at org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1202) ~[?:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1299) ~[?:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1197) ~[?:?]
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1610) ~[?:?]
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1546) ~[?:?]
	at org.objectweb.asm.MethodVisitor.visitMaxs(MethodVisitor.java:772) ~[?:?]
	at org.objectweb.asm.commons.LocalVariablesSorter.visitMaxs(LocalVariablesSorter.java:148) ~[?:?]
	at org.objectweb.asm.ClassReader.readCode(ClassReader.java:2636) ~[?:?]
	at org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1488) ~[?:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:718) ~[?:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:401) ~[?:?]
	at org.apache.aries.spifly.dynamic.ClientWeavingHook.weave(ClientWeavingHook.java:60) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.invokeWeavingHook(SecureAction.java:1204) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.transformClass(BundleWiringImpl.java:2357) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.findClass(BundleWiringImpl.java:2049) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1533) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:78) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1950) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:11) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:8) ~[?:?]
	at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:45) ~[?:?]
	at io.fabric8.kubernetes.client.osgi.ManagedKubernetesClient.<init>(ManagedKubernetesClient.java:137) ~[?:?]
	... 41 more
2021-11-24T16:03:48,909 | ERROR | FelixDispatchQueue | Framework                        | 60 - org.apache.aries.spifly.dynamic.bundle - 1.3.0 | FrameworkEvent ERROR
java.lang.IllegalArgumentException: Unsupported class file major version 61
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:196) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:177) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:163) ~[?:?]
	at org.objectweb.asm.ClassReader.<init>(ClassReader.java:284) ~[?:?]
	at org.apache.aries.spifly.dynamic.OSGiFriendlyClassWriter.getCommonSuperClass(OSGiFriendlyClassWriter.java:84) ~[?:?]
	at org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1202) ~[?:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1299) ~[?:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1197) ~[?:?]
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1610) ~[?:?]
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1546) ~[?:?]
	at org.objectweb.asm.MethodVisitor.visitMaxs(MethodVisitor.java:772) ~[?:?]
	at org.objectweb.asm.commons.LocalVariablesSorter.visitMaxs(LocalVariablesSorter.java:148) ~[?:?]
	at org.objectweb.asm.ClassReader.readCode(ClassReader.java:2636) ~[?:?]
	at org.objectweb.asm.ClassReader.readMethod(ClassReader.java:1488) ~[?:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:718) ~[?:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:401) ~[?:?]
	at org.apache.aries.spifly.dynamic.ClientWeavingHook.weave(ClientWeavingHook.java:60) ~[?:?]
	at org.apache.felix.framework.util.SecureAction.invokeWeavingHook(SecureAction.java:1204) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.transformClass(BundleWiringImpl.java:2357) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.findClass(BundleWiringImpl.java:2049) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1533) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:78) ~[?:?]
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1950) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:11) ~[?:?]
	at io.fabric8.kubernetes.client.ConfigBuilder.<init>(ConfigBuilder.java:8) ~[?:?]
	at io.fabric8.kubernetes.client.BaseClient.<init>(BaseClient.java:45) ~[?:?]
	at io.fabric8.kubernetes.client.osgi.ManagedKubernetesClient.<init>(ManagedKubernetesClient.java:137) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
	at org.apache.felix.scr.impl.inject.internal.ComponentConstructorImpl.newInstance(ComponentConstructorImpl.java:316) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:286) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:776) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437) ~[?:?]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:667) ~[?:?]
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:305) ~[?:?]
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:554) ~[?:?]
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:70) ~[?:?]
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:421) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49) ~[?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:420) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232) ~[osgi.core-7.0.0.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450) ~[osgi.core-7.0.0.jar:?]
	at org.apache.felix.framework.EventDispatcher.invokeBundleListenerCallback(EventDispatcher.java:915) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:834) ~[?:?]
	at org.apache.felix.framework.EventDispatcher.fireBundleEvent(EventDispatcher.java:516) ~[?:?]
	at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4817) ~[?:?]
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2336) ~[?:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998) ~[?:?]
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984) ~[?:?]
	at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.startBundle(BundleInstallSupportImpl.java:165) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.startBundle(FeaturesServiceImpl.java:1160) ~[?:?]
	at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:1041) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069) ~[?:?]
	at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
```

**Comment by manusa** [comment]
It seems that after the Karaf update (#3667) tests are finally passing in Java 17 :tada:

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java [hunk 1]
```diff
@@ -162,7 +162,7 @@ void testSerializeYamlWithAlias() {
       .isNotNull()
       .hasFieldOrPropertyWithValue("metadata.name", "test-pod-with-alias")
       .hasFieldOrPropertyWithValue("spec.nodeSelector.workload", "build")
-      .hasFieldOrPropertyWithValue("spec.tolerations.size", 1)
+      .returns(1, p -> p.getSpec().getTolerations().size())
```
Comment (rohanKanojia) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java:165]: why is this change required?

Comment (manusa) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/utils/SerializationTest.java:165]: I think that in Java 17 AssertJ can't use reflection to access/invoke the list size method.

**Review by rohanKanojia — APPROVED** [review]

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -40,7 +40,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        java: [8,11]
+        java: [8,17]
```
Comment (andreaTP) [inline .github/workflows/build.yml:43]: Have you considered dropping `8` instead of `11`?

Java 8 support already ended on 31 Mar 2022.

Comment (manusa) [reply .github/workflows/build.yml:43]: We want Fabric8 to be 8 compatible.

We will probably drop its support once everyone migrates to Spring 6 :wink:, or until we can no longer sustain it, or if there's some 11 specific feature that we need. For the last part, what we've followed is the creation of JDK11 specific optional modules, which seem to work well.

Comment (andreaTP) [reply .github/workflows/build.yml:43]: Fair enough 👍

**Review by andreaTP — APPROVED** [review]

**Comment by rohanKanojia** [comment]
What is status of this PR?

**Comment by manusa** [comment]
Needs to be worked on, I'll rebase later to check what might be needed

**Review by sunix — APPROVED** [review]

**Comment by manusa** [comment]
https://github.com/apache/karaf/pull/1493
